(DOCSP-9483): Intro for Web Devs (#180)
mongodb/docs-realm: source/get-started/introduction-web.txt

================================================
Introduction to MongoDB Realm for Web Developers
================================================

.. default-domain:: mongodb

> Modern web applications are more powerful than ever before and the data-driven

Reviewer: I really like this opening paragraph 🥇

approach popularized by frontend libraries like React and Angular has empowered
developers to create web apps that rival native applications in terms of

> performance and user experience. However, as the prevalance and usage of web

Reviewer: typo here - "prevalance" should be "prevalence"

applications grow, so do the demands of maintaining and delivering the data that
drives them.

In addition to frontend applications, developers typically must build secure
application servers that:

- :ref:`Authenticate and manage users <intro-web-users>`
- :ref:`Validate and secure data <intro-web-schema>`
- :ref:`Run custom logic and respond to requests <intro-web-functions>`
- :ref:`Expose data to client applications <intro-web-graphql>`

You can solve each of these problems in isolation with a wide variety of
libraries and frameworks, but deciding which solution solves each problem with
the right tradeoffs for your use case can be a daunting challenge. MongoDB Realm
is a managed application backend that streamlines these common challenges and

> allows you to shift your focus from wiring up a backend components to writing

Reviewer: Necessary Fix:
---------------
Typo: "MongoDB Realm is a managed application backend that streamlines these common challenges and allows you to shift your focus from wiring up a backend components to writing the code that makes your app great."
should not be "a backend components" it should be "backend components" 

so the full sentence should read: "MongoDB Realm is a managed application backend that streamlines these common challenges and allows you to shift your focus from wiring up backend components to writing the code that makes your app great."

Suggestion, non-blocking:
--------------------------
This might read better if we 
* changed "managed backend" to something else - what does this even mean to reader who don't know of stitch? It's currently a bit jargony 
* changed "wiring up backend components" - to me that's a little vague. An alternative might be: "writing tedious backend logic" 
* "makes your application code great" - gives me big MAGA vibes (this is just a personal :kirbysad: for me)

Here's an alternative sentence: "MongoDB Realm is a comprehensive platform that streamlines these common challenges by providing you with easy to use backend features so you can focus on writing the code that your users want."

> allows you to shift your focus from wiring up a backend components to writing

Reviewer 2: I would also suggest changing "wiring up" to "setting up" -- "wiring up" could be confusing to ESL readers.

the code that makes your app great.

.. _intro-web-users:

User Authentication and Management
----------------------------------
  
Secure user authentication is essential for most apps. User-driven applications
need to know the identity of each user in order to securely store their data and
make it available across sessions and devices. Some apps authenticate users with
traditional email/password combinations and others may use API keys or external
services like Facebook, Google, and Sign in with Apple. These methods can be

> complex and hard to correctly develop and integrate into your app.

Reviewer: Suggestion: remove "hard to correctly develop" its redundant since we've already said these methods are complex and "correctly develop" feels a bit judgy

> complex and hard to correctly develop and integrate into your app.

Reviewer 2: While I do agree that these services can be hard/complex to correctly develop and integrate into apps, I think an even better sell is the fact that it's a lot of manual boilerplate work in every application you develop. The "hard/complex" point might not matter to some devs (who embrace a good challenge!), but the idea of avoiding retreading the same ground in every application has universal appeal.

> complex and hard to correctly develop and integrate into your app.

Author: My goal here was to split both infinitives ("to develop" and "to integrate") but I see the ambiguity here..

Per Moe's suggestion, I think it's probably easier to just only have one infinitive and to drop the adverb.

> MongoDB Realm provides built-in user management and authentication methods, and

Author: Going with your alternative wording here which means we won't need this comma

> MongoDB Realm provides built-in user management and authentication methods, and

Author: Adding a fragment about the complexity/boilerplate per Nate's comment

"MongoDB Realm provides built-in user management and authentication methods that
encapsulate this complexity and make it easy to integrate third-party
authentication providers like social media platforms and enterprise identity
management systems."

makes it easy to integrate third-party authentication providers like social
media platforms and enterprise identity management systems. You can enable
authentication providers with a simple server-side configuration and log in from

> any client application using the Realm SDK.

Reviewer 2: I love how you claim that Realm "makes this easy"... and then actually back it up with a concrete explanation. This is a great way to balance the market-y speak with technical writing.

To learn more about authentication in MongoDB Realm, see :doc:`Users & Authentication </authentication>`.

.. _intro-web-schema:

Schema Validation and Data Access Rules
---------------------------------------

> Modern applications require data to be available in consistent types and formats

Reviewer: This might read better as "Modern applications require data to be available **using** consistent types and formats **ensuring** that all of the application’s components can understand and work with **the data**."

It is also a little confusing on what you mean by "work with" here. Maybe further explanation is needed

to ensure that all of the application's components can understand and work with

> it. It's also important to ensure that any given piece of data is only

Reviewer 2: You use "to ensure" twice in a row -- might be a good idea to change up the second occurrence so this doesn't feel repetitive.

Author: Modified the first instead ;)

accessible or modifiable by an authorized user, such as an administrator or the
user that the data describes.

MongoDB Realm automatically validates data against a schema that you define
using a superset of the JSON schema standard that supports the BSON types used

> by MongoDB.

Reviewer 2: On one hand, this sentence is very very accurate.

On the other hand, for someone who's unfamiliar with MongoDB... this is a really complex thought to process. Do you think we can simplify this explanation and let folks learn about the dirty details through a link to the schema docs instead? Additionally, this explanation omits the fact that you can also define your schema via developer mode... and if we simplify it, we don't have to worry about glossing over those details. I would suggest dropping the BSON reference at least -- for this audience, and at this level of detail, I don't think anybody is going to worry about JSON vs. BSON.

Author: Generally agreed except for dev mode. That's not (yet) a concept for web so the only approach is to use Realm Schema.

To learn more about defining a schema and access rules for your data, see
:ref:`collection rules <mongodb-rules>`.

.. _intro-web-functions:

Event-Driven Serverless Functions
---------------------------------

> Some applications require that certain logic must run on the app server. Usually

Reviewer 2: ```suggestion
Many applications run certain privileged logic on the app server. Usually
```

To me "require that ... must run" is a bit of a mouthful. Do you think we can streamline this at all?

this logic deals with sensitive user data or coordinates data from multiple
external services. Apps might also need to respond automatically in real time
when something happens in the app or in an external service.

MongoDB Realm allows you to define serverless JavaScript functions that you can
call from client applications or automatically execute in response to trigger

> events, such as an update to a document in MongoDB or a scheduled cron job. You

Reviewer 2: ```suggestion
events, such as an update to a document in MongoDB or at a scheduled time. You
```

While cron is pretty well known, I don't want anyone to get confused here if they haven't had the chance to learn about it yet. Also, could we link to the cron reference or scheduled triggers here?

can also write webhooks that execute a function in response to external
applications and services.

To learn more about defining and using serverless functions, see :doc:`Functions
</functions>`.

.. _intro-web-graphql:

Secure Client-Side Queries
--------------------------

> Modern web applications rely on data, so developers need to be able to access

Reviewer 2: +1 on this, since "need" shows up later in this same sentence.

the data that they need quickly and efficiently in the proper shape. However,
web browsers are insecure clients, so web applications typically store data on a
remote server and use a client-side API to access the data in order to prevent

> unauthorized access.

Reviewer 2: I think this is getting at the right points, but I think we can highlight some parts of the status quo for web apps to explain the advantages of MongoDB Data Access/GraphQL:
```
Modern web applications rely on data, so developers need to be able to access
the data that they need quickly and efficiently. However,
web browsers are insecure clients, so web applications
typically access persisted data on a remote server that
exposes an API to the client in order to prevent
unauthorized access. To enable this, developers must
implement an entire API, including a secure
authentication layer and methods to access individual
objects in the database.

```

Bringing up this "manually implementing an entire API" point makes the advantages of MongoDB Data Access and GraphQL really clear to me -- Realm basically provides the entire middleware layer for you. What do you think?

a declarative query language that's designed to empower developers to access
exactly the data they need and make data access and manipulation more efficient.

GraphQL has a couple of primary benefits for web developers:

- **Minimize Bandwidth**: GraphQL queries specify the exact fields to return,
  which allows you to avoid overfetching as you might from a static resource
  endpoint. You can specify the exact shape of the data your app needs and the
  GraphQL API returns exactly that data, no more and no less.

- **Minimize Roundtrips**: A single GraphQL endpoint handles all requests, which
  differs from other HTTP API architectures like REST, which assigns every
  resource to a separate URL. This means you can group all of your GraphqQL
  queries into a single request instead of sending a separate request to every
  resource endpoint.

MongoDB Realm automatically generates a GraphQL API for your application based
on the :ref:`data model <intro-web-schema>` that you define. The API includes a
set of CRUD queries and mutations that you can access through any standard
GraphQL client or directly over HTTP. If you need more than just the generated

> queries, you can define custom resolvers backed by :ref:`functions

Author: I considered it but in general I wanted to limit the outbound links on this page to the CTAs at the end of each section. The other links generally go to either (a) Wikipedia for additional context or (b) another section on the same page.

They'll learn more about custom resolvers if they follow the link at the end of the section, and I don't want them to churn off the page before that!

> queries, you can define custom resolvers backed by :ref:`functions

Reviewer 3: Sounds good to me.

<intro-web-functions>` and call them alongside your CRUD resolvers.

To learn more about the Realm GraphQL API, see :doc:`GraphQL </graphql>`.

Client-Side MongoDB Queries 
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Traditionally developers need to execute database queries on the server and wrap

> them behind another API layer (e.g. :wikipedia:`REST

Reviewer 3: I was told to avoid latin abbreviations, i.e. e.g.

<Representational_state_transfer>`) but this adds an additional layer of
indirection and lacks the flexibility of directly querying a database.

MongoDB Realm includes a built-in MongoDB service that allows you to use the
MongoDB query language, a powerful, idiomatic API that naturally fits the
MongoDB document model, anywhere in your application. You can query MongoDB
Atlas as if you were using a database driver and Realm will enforce :ref:`data
access rules <intro-web-schema>` for all requests.

To learn more about the built-in MongoDB service, see :doc:`MongoDB Atlas
</mongodb>`.

> Summary

Reviewer 3: Maybe before the summary, we have another section like "Get Started Now" that links to the page on creating a Realm App?

================================================
Introduction to MongoDB Realm for Web Developers
================================================

.. default-domain:: mongodb

Modern web applications are more powerful than ever before, and the data-driven
approach popularized by frontend libraries like React and Angular has empowered
developers to create web apps that approach the look and feel of native
applications. However, as the prevalence and usage of web applications grow, so
do the demands of maintaining and delivering the data that drives them.

In addition to frontend applications, developers typically must build secure
application servers that implement features like user authentication, data
validation, and business logic. These servers always need an API so that they
can communicate with browsers, so developers need to create HTTP routes that

> respond to incoming request. You can solve each of these problems in isolation

Reviewer: should this be "requests"? As in:
"These servers always need an API so that they can communicate with browsers, so developers need to create HTTP routes that respond to incoming **requests**."

unauthorized access. To do this, developers must implement an entire API that
provides access to objects in the database and includes a secure authentication
layer.

MongoDB Realm automatically generates a :ref:`GraphQL API <intro-web-graphql>`
and the Realm Web SDK includes a :ref:`secure MongoDB client <intro-web-mql>`
that uses standard MongoDB query language to access and modify documents in
Atlas. You can use either or both of these APIs to declaratively work with your
application's data.

Realm secures data with server-side rules that dynamically return and withhold
data depending on the user that issued a query. Rules consist of one or more
user roles for each collection. A role determines whether a given user has
permission to insert or modify a document as well as if they can read or write
specific fields in the document. Realm dynamically determines a role for each
document in every request and enforces the role's permissions for its
corresponding document. This ensures that users only read and modify documents
for which they have permission.

.. _intro-web-graphql:

GraphQL for MongoDB
~~~~~~~~~~~~~~~~~~~

> :graphql:`GraphQL <>` is a standard, open-source API architecture that is often

Reviewer: Often viewed by who? Would it better if we wrote it as:
"GraphQL is a standard, open-source API architecture that is an alternative to REST or other HTTP APIs." that way it avoids passive voice and is a little more declarative

viewed as an alternative to REST or other HTTP APIs. It features a declarative
query language that empowers developers to access exactly the data they need and
to make network requests for data more efficient.

GraphQL has a couple of primary benefits for web developers:

- **Minimize Bandwidth**: GraphQL queries specify the exact fields to return,

>   which allows you to avoid overfetching as you might from a static resource

Reviewer: should it be "over-fetching" rather than "overfetching"?

Author: I've seen both - the community likes hyphenated but the GraphQL docs don't hyphenate. I'll go with the community!
